File: LeetCode/LeetCode Practice Problems/890- Find and Replace Pattern.cpp
class Solution {
public:
    bool check(string s,string p)
    {
        map<char , char> m1,m2;
        for(int i=0;i<s.length();i++)
        {
            char c1 = s[i];
            char c2 = p[i];
            if(m1.find(c1) == m1.end()) m1[c1] = c2;
            if(m2.find(c2) == m2.end()) m2[c2] = c1;
            if(m1[c1] != c2) return false;
            if(m2[c2] != c1) return false;
        }
        return true;
    }
    
    vector<string> findAndReplacePattern(vector<string>& words, string pattern) {
        vector<string> ans;
        for(string s : words)
            if(check(s , pattern))
                ans.push_back(s);
        
        return ans;
    }
};

File: LeetCode/LeetCode Practice Problems/890-Find and Replace Pattern.java
class Solution {
    public List<String> findAndReplacePattern(String[] words, String pattern) {
        List<String> ans = new ArrayList();
        
        for(String word: words) {
            HashMap<Character,Character> map1 = new HashMap();
            HashMap<Character,Character> map2 = new HashMap();
            boolean flag = true;
            
            for(int i=0;i<word.length();i++) {
                char ch1 = word.charAt(i);
                char ch2 = pattern.charAt(i);
                
                if((map1.containsKey(ch1) && map1.get(ch1) != ch2) || (map2.containsKey(ch2) && map2.get(ch2) != ch1)) {
                    flag = false;
                    break;
                }
                
                map1.put(ch1 , ch2);
                map2.put(ch2 , ch1);
            }
            
            if(flag) ans.add(word);
        }
        
        return ans;
    }
}
